# .Business understanding

Aircraft vary widely in their safety records depending on manufacturer, model, maintenance history, usage, and geography.While modern aircraft are statistically very safe, historical data shows that certain models and types have accidents rates often linked to operational environment, mechanical complexities or outdated systems. a good safety record of an airline is critical towards its operational market, survival, reputation, prestige and most importantly passengers confidence towards its service offered. 

The company being new to aircraft enterprise, lacks experience in aircraft risk assessment. its therefore important to determine which aircraft has lowest risks for the organization to makes decisions on which aircraft to purchase. 

## .Business problem
The organization is expanding into the aviation industry and seeks to identify aircrafts with low-risks in order to proceed with purchase and starts its operations. the goal is to ensure safety and purchasing a reliable aircraft for commercial and private use.
### .Objectives and business questions
1) To identify  the safest aircrafts make/models

2) To evaluate aircraft safety Risks based on purpose of flight

3) To analyze regional operational risks 
 

## .Metric of success
The success of the business depends on evidence-based recommendations on low risk aircraft before initial purchase. insights must be actionable, understandable to non-technical leadership and can be used in procurement process.

# .Data understanding 
The data to be used in this analysis is from the National Transportation Safety Board that includes aviation accident data from 1962 to 2023 about civil aviation accidents and selected incidents in the United States and international waters.



## Loading dataset and getting the information

In [1]:
# importing necessary libraries to be used
import pandas as pd
import numpy as np
import seaborn as sns
from matplotlib import pyplot as plt

%matplotlib inline

In [2]:
# reading and loading the dataset
aviation_data = pd.read_csv("data\AviationData.csv", encoding = "latin1")
aviation_data.head()

C:\Users\User\anaconda3\envs\learn-env\lib\site-packages\IPython\core\interactiveshell.py:3145: DtypeWarning: Columns (6,7,28) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country Latitude Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States      NaN       NaN          NaN   
1   BRIDGEPORT, CA  United States      NaN       NaN          NaN   
2    Saltville, VA  United States  36.9222  -81.8781          NaN   
3       EUREKA, CA  United States      NaN       NaN          NaN   
4       Canton, OH  United States      NaN       NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

In [3]:
aviation_data.tail()

Event.Id Investigation.Type Accident.Number  Event.Date  \
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

            Location        Country Latitude Longitude Airport.Code  \
88884  Annapolis, MD  United States      NaN       NaN          NaN   
88885    Hampton, NH  United States      NaN       NaN          NaN   
88886     Payson, AZ  United States  341525N  1112021W          PAN   
88887     Morgan, UT  United States      NaN       NaN          NaN   
88888     Athens, GA  United States      NaN       NaN          NaN   

      Airport.Name  ... Purpose.of.flight         Air.carrier  \
88884          NaN  ...          Personal                 NaN   
88885          NaN  ...               NaN                 NaN   
88886       PAYSON  ...          Personal                 NaN   
88887          NaN  ...          Personal  MC CESSNA 210N LLC   
88888          NaN  ...          Personal                 NaN   

      Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
88884                  0.0                    1.0                  0.0   
88885                  0.0                    0.0                  0.0   
88886                  0.0                    0.0                  0.0   
88887                  0.0                    0.0                  0.0   
88888                  0.0                    1.0                  0.0   

      Total.Uninjured Weather.Condition  Broad.phase.of.flight Report.Status  \
88884             0.0               NaN                    NaN           NaN   
88885             0.0               NaN                    NaN           NaN   
88886             1.0               VMC                    NaN           NaN   
88887             0.0               NaN                    NaN           NaN   
88888             1.0               NaN                    NaN           NaN   

      Publication.Date  
88884       29-12-2022  
88885              NaN  
88886       27-12-2022  
88887              NaN  
88888       30-12-2022  

[5 rows x 31 columns]

In [4]:
aviation_data.sample(5, random_state= 4)

Event.Id Investigation.Type Accident.Number  Event.Date  \
38221  20001207X04511           Accident      LAX95FA321  1995-09-02   
33359  20001211X12901           Accident      DEN93LA088  1993-07-27   
65184  20080820X01268           Accident      CHI08WA258  2008-08-17   
82316  20180915X11112           Accident      WPR18LA262  2018-09-15   
87501  20220307104735           Accident      ANC22LA022  2022-03-05   

                        Location         Country Latitude Longitude  \
38221                PHOENIX, AZ   United States      NaN       NaN   
33359                 MEEKER, CO   United States      NaN       NaN   
65184  Covington, United Kingdom  United Kingdom      NaN       NaN   
82316              St. Johns, AZ   United States  343049N  1092213W   
87501               Newhalen, AK   United States  593854N  0154597W   

      Airport.Code                 Airport.Name  ... Purpose.of.flight  \
38221          DVT     PHOENIX DEER VALLEY MUNI  ...          Personal   
33359          NaN                          NaN  ...          Personal   
65184          NaN                          NaN  ...           Unknown   
82316          SJN  St Johns Industrial Airport  ...          Personal   
87501          NaN                          NaN  ...          Business   

                       Air.carrier Total.Fatal.Injuries  \
38221                          NaN                  2.0   
33359                          NaN                  0.0   
65184  Reconnaissance Ventures Ltd                  5.0   
82316                          NaN                  0.0   
87501                   SEND NORTH                  0.0   

      Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
38221                    0.0                  0.0             0.0   
33359                    1.0                  1.0             0.0   
65184                    0.0                  0.0             0.0   
82316                    0.0                  3.0             0.0   
87501                    1.0                  4.0             0.0   

      Weather.Condition  Broad.phase.of.flight  \
38221               VMC               Approach   
33359               VMC            Maneuvering   
65184               NaN                    NaN   
82316               VMC                    NaN   
87501               IMC                    NaN   

                                           Report.Status Publication.Date  
38221                                     Probable Cause       04-02-1997  
33359                                     Probable Cause       13-09-1994  
65184                                                NaN       25-09-2020  
82316  The pilot's inadequate preflight planning befo...       25-09-2020  
87501  The pilot’s decision to continue visual flight...       01-09-2022  

[5 rows x 31 columns]

In [5]:
# checking rows and columns shape
aviation_data.shape

(88889, 31)

The above dataset has a total of 88889 records and 31 features

In [6]:
# column names
aviation_data.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

In [7]:
# checking data info, dtypes and non null counts
aviation_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50249 non-null  object 
 9   Airport.Name            52790 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87572 non-null  object 
 14  Make                    88826 non-null

From the above information, its clearly notable that the dataset has a total of 5 columns with numerical data types and
26 columns consisting of categorical data.some columns such as Event.Date have data type object instead of date time, most of the variables have null values only a few variables have no null values
that indicates that data needs to cleaned before analysis.

In [8]:
# checking for statistical info
aviation_data.describe().T

count      mean        std  min  25%  50%  75%  \
Number.of.Engines       82805.0  1.146585   0.446510  0.0  1.0  1.0  1.0   
Total.Fatal.Injuries    77488.0  0.647855   5.485960  0.0  0.0  0.0  0.0   
Total.Serious.Injuries  76379.0  0.279881   1.544084  0.0  0.0  0.0  0.0   
Total.Minor.Injuries    76956.0  0.357061   2.235625  0.0  0.0  0.0  0.0   
Total.Uninjured         82977.0  5.325440  27.913634  0.0  0.0  1.0  2.0   

                          max  
Number.of.Engines         8.0  
Total.Fatal.Injuries    349.0  
Total.Serious.Injuries  161.0  
Total.Minor.Injuries    380.0  
Total.Uninjured         699.0

In [9]:
# categorical info
aviation_data.describe(include= "O").head(10)

Event.Id Investigation.Type Accident.Number  Event.Date  \
count            88889              88889           88889       88889   
unique           87951                  2           88863       14782   
top     20001214X45071           Accident      DCA22WA167  2000-07-08   
freq                 3              85015               2          25   

             Location        Country Latitude Longitude Airport.Code  \
count           88837          88663    34382     34373        50249   
unique          27758            219    25592     27156        10375   
top     ANCHORAGE, AK  United States  332739N  0112457W         NONE   
freq              434          82248       19        24         1488   

       Airport.Name  ... Amateur.Built    Engine.Type FAR.Description  \
count         52790  ...         88787          81812           32023   
unique        24871  ...             2             13              31   
top         Private  ...            No  Reciprocating             091   
freq            240  ...         80312          69530           18221   

       Schedule Purpose.of.flight Air.carrier Weather.Condition  \
count     12582             82697       16648             84397   
unique        3                26       13590                 4   
top        NSCH          Personal       Pilot               VMC   
freq       4474             49448         258             77303   

       Broad.phase.of.flight   Report.Status Publication.Date  
count                  61724           82508            75118  
unique                    12           17075             2924  
top                  Landing  Probable Cause       25-09-2020  
freq                   15428           61754            17019  

[4 rows x 26 columns]

In [292]:
# us_code_data= pd.read_csv("data/USState_Codes.csv", encoding= "latin1")

## Data cleaning

In [23]:
# creating a copy of a data frame to be used in cleaning
aviation_cleaned_data = aviation_data.copy(deep= True)

In [24]:
# checking the Country column
aviation_cleaned_data['Country'].value_counts(dropna= False)

United States     82248
Brazil              374
Canada              359
Mexico              358
United Kingdom      344
                  ...  
AY                    1
Ivory Coast           1
Nauru                 1
Libya                 1
Chad                  1
Name: Country, Length: 220, dtype: int64

Observation
from the above information, we see most of the data comes from the United States. so we will filter out the other
countries to remain with united states to be used in this analysis

In [70]:
# filtering dataset to include country USA
us_aviation_accidents_data= aviation_cleaned_data[aviation_cleaned_data['Country']== "United States"]
us_aviation_accidents_data

Event.Id Investigation.Type Accident.Number  Event.Date  \
0      20001218X45444           Accident      SEA87LA080  1948-10-24   
1      20001218X45447           Accident      LAX94LA336  1962-07-19   
2      20061025X01555           Accident      NYC07LA005  1974-08-30   
3      20001218X45448           Accident      LAX96LA321  1977-06-19   
4      20041105X01764           Accident      CHI79FA064  1979-08-02   
...               ...                ...             ...         ...   
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

              Location        Country Latitude Longitude Airport.Code  \
0      MOOSE CREEK, ID  United States      NaN       NaN          NaN   
1       BRIDGEPORT, CA  United States      NaN       NaN          NaN   
2        Saltville, VA  United States  36.9222  -81.8781          NaN   
3           EUREKA, CA  United States      NaN       NaN          NaN   
4           Canton, OH  United States      NaN       NaN          NaN   
...                ...            ...      ...       ...          ...   
88884    Annapolis, MD  United States      NaN       NaN          NaN   
88885      Hampton, NH  United States      NaN       NaN          NaN   
88886       Payson, AZ  United States  341525N  1112021W          PAN   
88887       Morgan, UT  United States      NaN       NaN          NaN   
88888       Athens, GA  United States      NaN       NaN          NaN   

      Airport.Name  ... Purpose.of.flight         Air.carrier  \
0              NaN  ...          Personal                 NaN   
1              NaN  ...          Personal                 NaN   
2              NaN  ...          Personal                 NaN   
3              NaN  ...          Personal                 NaN   
4              NaN  ...          Personal                 NaN   
...            ...  ...               ...                 ...   
88884          NaN  ...          Personal                 NaN   
88885          NaN  ...               NaN                 NaN   
88886       PAYSON  ...          Personal                 NaN   
88887          NaN  ...          Personal  MC CESSNA 210N LLC   
88888          NaN  ...          Personal                 NaN   

      Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
0                      2.0                    0.0                  0.0   
1                      4.0                    0.0                  0.0   
2                      3.0                    NaN                  NaN   
3                      2.0                    0.0                  0.0   
4                      1.0                    2.0                  NaN   
...                    ...                    ...                  ...   
88884                  0.0                    1.0                  0.0   
88885                  0.0                    0.0                  0.0   
88886                  0.0                    0.0                  0.0   
88887                  0.0                    0.0                  0.0   
88888                  0.0                    1.0                  0.0   

      Total.Uninjured Weather.Condition  Broad.phase.of.flight  \
0                 0.0               UNK                 Cruise   
1                 0.0               UNK                Unknown   
2                 NaN               IMC                 Cruise   
3                 0.0               IMC                 Cruise   
4                 0.0               VMC               Approach   
...               ...               ...                    ...   
88884             0.0               NaN                    NaN   
88885             0.0               NaN                    NaN   
88886  

In [71]:
us_aviation_accidents_data = us_aviation_accidents_data.copy()

In [72]:
# checking for stat summary of numerical columns before filling in NaN values
us_aviation_accidents_data.describe().T

count      mean        std  min  25%  50%  75%  \
Number.of.Engines       80373.0  1.135481   0.427286  0.0  1.0  1.0  1.0   
Total.Fatal.Injuries    71594.0  0.421683   2.433647  0.0  0.0  0.0  0.0   
Total.Serious.Injuries  70873.0  0.257178   1.144189  0.0  0.0  0.0  0.0   
Total.Minor.Injuries    71519.0  0.332974   1.306604  0.0  0.0  0.0  0.0   
Total.Uninjured         77243.0  4.302448  23.794728  0.0  0.0  1.0  2.0   

                          max  
Number.of.Engines         8.0  
Total.Fatal.Injuries    265.0  
Total.Serious.Injuries  137.0  
Total.Minor.Injuries    125.0  
Total.Uninjured         699.0

In [73]:
# filling missing numerical injury columns with 0
injury_columns= ['Total.Fatal.Injuries', 'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured']
us_aviation_accidents_data[injury_columns] = us_aviation_accidents_data[injury_columns].fillna(0)

In [74]:
# checking null values
us_aviation_accidents_data[injury_columns].isna().sum()

Total.Fatal.Injuries      0
Total.Serious.Injuries    0
Total.Minor.Injuries      0
Total.Uninjured           0
dtype: int64

In [75]:
 # convert injury columns to numeric
for col in injury_columns:
    us_aviation_accidents_data[col]= pd.to_numeric(us_aviation_accidents_data[col], errors= "coerce")

Cleaning categorical columns

In [76]:
# clean categorical columns
categorical_columns= ['Aircraft.Category', 'Make', 'Model', 'Purpose.of.flight', 'Injury.Severity', 
                      'Aircraft.damage', 'Weather.Condition']

for col in us_aviation_accidents_data.columns:
    if us_aviation_accidents_data[col].dtype == "O":
        us_aviation_accidents_data[col] = us_aviation_accidents_data[col].str.strip().str.lower()

In [ ]:
# checking categorical columns
us_aviation_accidents_data[categorical_columns].isna().sum()

In [ ]:
# filling in Aircraft.Category 
us_aviation_accidents_data['Aircraft.Category'].value_counts(dropna= False)

The above Column has 54094 misiing values and its notable that some values are abbreviation of others for instance
wsft is used instead of weight-shift and ultr inplace of ultralight

In [77]:
# creating a dic of category abbreviations
category_abb = {'wsft': 'weight-shift',
               'ultr': 'ultralight',
                'unk': 'unknown'}

In [78]:
# replacing abbreviations
us_aviation_accidents_data['Aircraft.Category'] = us_aviation_accidents_data['Aircraft.Category'].replace(category_abb)

In [79]:
# checking
us_aviation_accidents_data['Aircraft.Category'].value_counts()

airplane             24229
helicopter            2723
glider                 503
balloon                229
gyrocraft              172
weight-shift           170
powered parachute       90
ultralight              26
unknown                  4
blimp                    4
powered-lift             3
rocket                   1
Name: Aircraft.Category, dtype: int64

In [80]:
# filling in null values with unknown
us_aviation_accidents_data['Aircraft.Category']= us_aviation_accidents_data['Aircraft.Category'].fillna('unknown')

In [81]:
# checking
us_aviation_accidents_data['Aircraft.Category'].isna().sum()

0

In [82]:
# checking null values
# filling in make column

us_aviation_accidents_data['Make'].isna().sum()

21

In [83]:
# filling in missing values with unknown
us_aviation_accidents_data['Make']= us_aviation_accidents_data['Make'].fillna("unknown")

In [84]:
# checking
us_aviation_accidents_data['Make'].isna().sum()

0

In [85]:
# cleaning model column
us_aviation_accidents_data['Model'].isna().sum()

38

In [86]:
# checking value counts
us_aviation_accidents_data[['Model']].value_counts()

Model             
152                   2323
172                   1637
172n                  1136
pa-28-140              910
150                    790
                      ... 
lg2h                     1
libelle                  1
liberty xl               1
lighthizer special       1
&gcbc                    1
Length: 10786, dtype: int64

In [87]:
# filling in null values with unkown
us_aviation_accidents_data['Model']= us_aviation_accidents_data['Model'].fillna("unknown")

In [90]:
# cleaning purpose of flight column
us_aviation_accidents_data['Purpose.of.flight'].isna().sum()

2429

In [92]:
# checking for value counts
us_aviation_accidents_data['Purpose.of.flight'].value_counts().head()

personal              48544
instructional         10429
unknown                5739
aerial application     4627
business               3843
Name: Purpose.of.flight, dtype: int64

In [93]:
# filling in missing values with unknown
us_aviation_accidents_data['Purpose.of.flight']= us_aviation_accidents_data['Purpose.of.flight'].fillna('unknown')

In [95]:
# checking effectiveness
# us_aviation_accidents_data.info()

In [112]:
# cleaning Injury.Severity
us_aviation_accidents_data['Injury.Severity'].isna().sum()

108

In [113]:
us_aviation_accidents_data['Injury.Severity']= us_aviation_accidents_data['Injury.Severity'].fillna('unknown')

In [116]:
# cleaning location column
us_aviation_accidents_data['Location'].value_counts(dropna= True)

anchorage, ak        548
miami, fl            275
houston, tx          271
albuquerque, nm      265
chicago, il          256
                    ... 
diamond, or            1
near honolulu, hi      1
pisgah, ia             1
walton, ks             1
santaquin, ut          1
Name: Location, Length: 17588, dtype: int64

In [121]:
# extracting states from location
us_aviation_accidents_data['State']= aviation_cleaned_data['Location'].str.split(",").str[-1].str.strip()

In [124]:
# checking
us_aviation_accidents_data['State'].isna().sum()

11

In [125]:
# filling missing values with unknown
us_aviation_accidents_data['State']= us_aviation_accidents_data['State'].fillna("unknown")

In [126]:
# checking columns
us_aviation_accidents_data.isna().sum()

Event.Id                      0
Investigation.Type            0
Accident.Number               0
Event.Date                    0
Location                     11
Country                       0
Latitude                  49992
Longitude                 50002
Airport.Code              33059
Airport.Name              30594
Injury.Severity               0
Aircraft.damage            1979
Aircraft.Category             0
Registration.Number          51
Make                          0
Model                         0
Amateur.Built                21
Number.of.Engines          1875
Engine.Type                3023
FAR.Description           54023
Schedule                  71951
Purpose.of.flight             0
Air.carrier               67753
Total.Fatal.Injuries          0
Total.Serious.Injuries        0
Total.Minor.Injuries          0
Total.Uninjured               0
Weather.Condition           645
Broad.phase.of.flight     21090
Report.Status              2611
Publication.Date          12681
State   

In [128]:
# cleaning weather column
us_aviation_accidents_data['Weather.Condition'].value_counts()

vmc    75317
imc     5618
unk      668
Name: Weather.Condition, dtype: int64

In [129]:
# filling in missing values with unknown
us_aviation_accidents_data['Weather.Condition']= us_aviation_accidents_data['Weather.Condition'].fillna('unk')

Cleaning location column

In [492]:
# checking for missing values
us_aviation_accidents_data['Location'].isna().sum()

11

In [493]:
# dropping missing rows in Location
us_aviation_accidents_data= us_aviation_accidents_data.dropna(subset= ['Location'])

In [494]:
us_aviation_accidents_data['Location'].isna().sum()

0

Cleaning Make Column

In [495]:
# checking null values


21

In [496]:
# value counts
us_aviation_accidents_data['Make'].value_counts(dropna= False)

Cessna                            21570
Piper                             11661
CESSNA                             4280
Beech                              4166
PIPER                              2507
                                  ...  
Darrell                               1
Wheeler Acft. Co.                     1
AIRMAX CONSTRUCOES AERONAUTICA        1
T BIRD GOLDEN CIRCLE AIR INC          1
Ragle                                 1
Name: Make, Length: 8002, dtype: int64

In [497]:
# lowering the make case to normalize or standardize the make
us_aviation_accidents_data['Make']= us_aviation_accidents_data['Make'].str.strip().str.lower()

In [209]:
# making 
make_missing_model_known = (us_aviation_accidents_data['Make'].isnull()) & (us_aviation_cle_data['model'].notnull())
aviation_cleaned_data['make'].isnull().sum()

63

In [210]:
aviation_cleaned_data.loc[make_missing_model_known, 'make'] = "cessna"

# aviation_cleaned_data.loc[make_missing_model_known, 'make']

In [211]:
# checking null values in make column
aviation_cleaned_data['make'].isnull().sum()

43

In [212]:
# filling in missing make values with unkwnown
aviation_cleaned_data['make'] = aviation_cleaned_data['make'].fillna('unknown')

In [213]:
aviation_cleaned_data['make'].isnull().sum()

0

In [214]:
aviation_cleaned_data['make'].unique()

array(['stinson', 'piper', 'cessna', ..., 'james r dernovsek',
       'orlican s r o', 'royse ralph l'], dtype=object)

In [215]:
common_five= list(aviation_cleaned_data['make'].value_counts().head().index)
common_five

['cessna', 'piper', 'beech', 'boeing', 'bell']

cleaning model rows

In [216]:
# checking missing values
aviation_cleaned_data['model'].value_counts().head()

152          2367
172          1756
172N         1164
PA-28-140     932
150           829
Name: model, dtype: int64

Observation

The model column has 92 missing records

In [217]:
# checking the value counts of the most common model by make
aviation_cleaned_data[['make', 'model']].value_counts().head(30)

make     model     
cessna   152           2366
         172           1753
         172N          1163
piper    PA-28-140      932
cessna   150            829
         172M           798
         172P           689
         182            659
         180            621
         150M           585
piper    PA-18          578
         PA-28-180      572
         PA-18-150      571
         PA-28-161      565
         PA-28-181      529
bell     206B           516
boeing   737            489
piper    PA-38-112      468
cessna   150L           460
beech    A36            419
cessna   140            401
         170B           389
         172S           371
grumman  G-164A         365
piper    PA-32-300      356
cessna   182P           355
piper    PA-24-250      352
         PA-28R-200     334
         PA-12          324
cessna   A188B          318
dtype: int64

Observation

In cessna make, model 152, is the most common model, and in piper make the most common model is PA-28-140,
in bell make is 206B, boeing is 737, so we are going to use the common make to fill in the missing values in model
using the the most common model, i.e if the make row is cessna and the model row is missing we are going to fill it with
the most common model of cessna and so on with other model rows


In [218]:
# creating a list of most common 30 make 
valid_makes= ['cessna', 'piper', 'beech', 'boeing', 'bell', 'mooney', 'robinson', 
              'grumman', 'bellanca', 'hughes', 'schweizer', 'air tractor', 'aeronca', 'mcdonnell douglas', 'maule', 
              'champion', 'stinson']

In [219]:
# filtering the dataset to only valid makes
aviation_data_filtered = aviation_cleaned_data[(aviation_cleaned_data['make'].isin(valid_makes)) & 
                                               (aviation_cleaned_data['model'].notna())]


In [220]:
# checking rows where model is missing
missing_model_aviation = aviation_cleaned_data[aviation_cleaned_data['model'].isna()]


In [221]:
# creating  make to model dict
model_mapping= (aviation_data_filtered.groupby('make')['model'].
               agg(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan).to_dict()
               )

In [222]:
# filling missing models row by row
for index, row in missing_model_aviation.iterrows():
    make= row['make']
    if make in model_mapping:
        aviation_cleaned_data.at[index, 'model']= model_mapping[make]

In [223]:
# checking missing values in model
aviation_cleaned_data['model'].isna().sum()

78

In [224]:
aviation_cleaned_data['model'] = aviation_cleaned_data['model'].fillna('unknown')

In [225]:
aviation_cleaned_data['model'].isna().sum()

0

cleaning aircraft category records

In [226]:
# checking all unique values
aircraft_category = aviation_cleaned_data['aircraft category'].value_counts(dropna= False)
aircraft_category

NaN                  56602
Airplane             27617
Helicopter            3440
Glider                 508
Balloon                231
Gyrocraft              173
Weight-Shift           161
Powered Parachute       91
Ultralight              30
Unknown                 14
WSFT                     9
Powered-Lift             5
Blimp                    4
UNK                      2
Rocket                   1
ULTR                     1
Name: aircraft category, dtype: int64

In [227]:
aviation_cleaned_data['aircraft category'].unique()

array([nan, 'Airplane', 'Helicopter', 'Glider', 'Balloon', 'Gyrocraft',
       'Ultralight', 'Unknown', 'Blimp', 'Powered-Lift', 'Weight-Shift',
       'Powered Parachute', 'Rocket', 'WSFT', 'UNK', 'ULTR'], dtype=object)

Observation some unique values are abbreviation of the full values such as WSFT simply means 
Weight-Shift category, ULTR means Ultralight aircraft category and UNK means Unknown

In [228]:
# changing the abbreviations to full values
# creating a dic of category abbreviations
category_abb = {'WSFT': 'Weight-Shift',
               'ULTR': 'Ultralight',
               'UNK': 'Unknown'}

In [42]:
# replacing abbreviations
aviation_cleaned_data['aircraft category'] = aviation_cleaned_data['aircraft category'].replace(category_abb)

KeyError: 'aircraft category'

In [230]:
# checking unique values after replacing the abbreviations
aviation_cleaned_data['aircraft category'].unique()

array([nan, 'Airplane', 'Helicopter', 'Glider', 'Balloon', 'Gyrocraft',
       'Ultralight', 'Unknown', 'Blimp', 'Powered-Lift', 'Weight-Shift',
       'Powered Parachute', 'Rocket'], dtype=object)

In [231]:
# checking null values
aviation_cleaned_data['aircraft category'].isna().sum()

56602

observation 
There are 56602 aircraft category missing records in dataset.

In [250]:
# filling in missing values
# creating a df where aircraft category is missing

aircraft_category_missing_df = aviation_cleaned_data[aviation_cleaned_data['aircraft category'].isna()]
# aircraft_category_missing_df

In [251]:
# creating a key using both make and model to fill the missing rows where make and model are not missing and aircraft
# category is misiing
make_model_category = (aviation_cleaned_data
                       .groupby(['make', 'model'])['aircraft category']
                       .agg(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)
                       .to_dict()
                      )

In [252]:
# creating a loop to fill using (make model)
for index, row in aircraft_category_missing_df.iterrows():
    key= (row['make'], row['model'])
    if key in make_model_category:
        aviation_cleaned_data.at[index, "aircraft category"]= make_model_category[key]

In [ ]:
aviation_cleaned_data['aircraft category'].isna().sum()

In [253]:
aviation_cleaned_data['aircraft category'] = aviation_cleaned_data['aircraft category'].fillna('unknown')

In [254]:
aviation_cleaned_data['aircraft category'].isna().sum()

0

cleaning injury severity column

In [255]:
# checking columns value counts
unique_vals = aviation_cleaned_data['injury severity'].unique()

In [256]:
# standardizing injury severity values
print(type(unique_vals))

<class 'numpy.ndarray'>


In [257]:
aviation_cleaned_data.columns

Index(['event id', 'investigation type', 'accident number', 'event date',
       'location', 'country', 'airport code', 'airport name',
       'injury severity', 'aircraft damage', 'aircraft category',
       'registration number', 'make', 'model', 'amateur built',
       'number of engines', 'engine type', 'purpose of flight',
       'total fatal injuries', 'total serious injuries',
       'total minor injuries', 'total uninjured', 'weather condition',
       'broad phase of flight', 'report status', 'publication date'],
      dtype='object')

In [258]:
aviation_cleaned_data['purpose of flight'].value_counts(dropna= False)

Personal                     49448
Unknown                      12994
Instructional                10601
Aerial Application            4712
Business                      4018
Positioning                   1646
Other Work Use                1264
Ferry                          812
Aerial Observation             794
Public Aircraft                720
Executive corporate            553
Flight Test                    405
Skydiving                      182
Air Race show                  158
External Load                  123
Public Aircraft - Federal      105
Banner Tow                     101
Public Aircraft - Local         74
Public Aircraft - State         64
Glider Tow                      53
Firefighting                    40
Air Drop                        11
ASHO                             6
PUBS                             4
PUBL                             1
Name: purpose of flight, dtype: int64

In [259]:
aviation_cleaned_data['purpose of flight'] = aviation_cleaned_data['purpose of flight'].str.replace("/", " ")

In [260]:
aviation_cleaned_data['purpose of flight'].value_counts(dropna= False)

Personal                     49448
Unknown                      12994
Instructional                10601
Aerial Application            4712
Business                      4018
Positioning                   1646
Other Work Use                1264
Ferry                          812
Aerial Observation             794
Public Aircraft                720
Executive corporate            553
Flight Test                    405
Skydiving                      182
Air Race show                  158
External Load                  123
Public Aircraft - Federal      105
Banner Tow                     101
Public Aircraft - Local         74
Public Aircraft - State         64
Glider Tow                      53
Firefighting                    40
Air Drop                        11
ASHO                             6
PUBS                             4
PUBL                             1
Name: purpose of flight, dtype: int64

In [261]:
aviation_cleaned_data['purpose of flight']= aviation_cleaned_data['purpose of flight'].fillna("Unknown")

In [262]:
aviation_cleaned_data['purpose of flight'].isna().sum()

0

ANALYSIS

In [263]:
aviation_cleaned_data["make"].sort_values(by= "purpose of flight", order= max)

TypeError: sort_values() got an unexpected keyword argument 'by'

In [264]:
# Creating injury columns
injury_cols= ['total uninjured', 'total minor injuries', 'total serious injuries', 'total fatal injuries']

# grouping by make and model
# common_five= aviation_cleaned_data['make'].value_counts(dropna= False).head()
aviation_cleaned_data.groupby(["make", "model"])[injury_cols].sum().sort_values(by= "total uninjured", ascending= False).head()

total uninjured  total minor injuries  \
make              model                                             
boeing            737                 25448                    91   
                  777                  9419                    23   
                  767                  6365                    30   
                  747-400              6260                    41   
mcdonnell douglas DC-10-10             6248                   280   

                            total serious injuries  total fatal injuries  
make              model                                                   
boeing            737                          388                  1348  
                  777                           27                     0  
                  767                          111                   128  
                  747-400                       43                    83  
mcdonnell douglas DC-10-10                      67                   111

In [298]:
# checking for duplicates in columns


NameError: name 'us_aviation_accident_data' is not defined

The above dataset has no duplicated records

In [270]:
aviation_cleaned_data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 26 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   event id                88889 non-null  object        
 1   investigation type      88889 non-null  object        
 2   accident number         88889 non-null  object        
 3   event date              88889 non-null  datetime64[ns]
 4   location                88837 non-null  object        
 5   country                 88663 non-null  object        
 6   airport code            50249 non-null  object        
 7   airport name            52790 non-null  object        
 8   injury severity         87889 non-null  object        
 9   aircraft damage         88889 non-null  object        
 10  aircraft category       88889 non-null  object        
 11  registration number     87572 non-null  object        
 12  make                    88889 non-null  object

In [268]:
aviation_cleaned_data['aircraft damage']= aviation_cleaned_data['aircraft damage'].fillna("Unknown")

In [269]:
aviation_cleaned_data['aircraft damage'].isna().sum()

0

In [273]:
aviation_cleaned_data['registration number'].value_counts(dropna= False)

NaN       1317
NONE       344
UNREG      126
None        65
UNK         13
          ... 
N1475M       1
N5874P       1
N124DK       1
N6408W       1
N8605T       1
Name: registration number, Length: 79106, dtype: int64

In [274]:
aviation_cleaned_data['country'].value_counts(dropna= True)

United States        82248
Brazil                 374
Canada                 359
Mexico                 358
United Kingdom         344
                     ...  
Montserrat               1
Albania                  1
Mauritania               1
Wallis and Futuna        1
Benin                    1
Name: country, Length: 219, dtype: int64